### PR TITLE
DS3: Convert post_fill to stage_post_fill for better performance

### PR DESCRIPTION
**Pull request by Mysteryem**
## What is this fixing or adding?

For the item smoothing feature, DS3 was iterating `MultiWorld.get_spheres()` once for every DS3 player in the multiworld in `post_fill`. By converting `post_fill` to `stage_post_fill`, `MultiWorld.get_spheres()` only needs to be iterated once, regardless of how many DS3 players are in the multiworld.

Additionally, if there are no players with an item smoothing option enabled, `stage_post_fill` now returns immediately without iterating `MultiWorld.get_spheres()` at all.

## How was this tested?

Fixed seeds were generated before and after this patch and seen to be identical.

With a multiworld containing 4 template DS3 yamls, generation duration (with `--skip_output`) decreased from ~0.42s to ~0.31s.

### CODE REVIEW

**Comment by Exempt-Medic** [comment]
THANK YOU. I tried doing this and gave up

**Comment by Exempt-Medic** [comment]
@nex3

File: worlds/dark_souls_3/__init__.py [hunk 1]
```diff
@@ -1375,106 +1376,125 @@ def post_fill(self):
...
+                    location.player != ds3_world.player or ds3_world._is_location_available(location)
+                ):
+                    full_items_by_name[location.item.name].append(location.item)
+
+            def smooth_items(item_order: List[Union[DS3ItemData, DarkSouls3Item]]) -> None:
+                """Rearrange all items in item_order to match that order.
+
+                Note: this requires that item_order exactly matches the number of placed items from this
+                world matching the given names.
+                """
```
Comment (Exempt-Medic) [inline worlds/dark_souls_3/__init__.py:1426]: Added a space in the changes
```suggestion
                """Rearrange all items in item_order to match that order.
                Note: this requires that item_order exactly matches the number of placed items from this
                world matching the given names.
                """
```

Comment (Mysteryem) [reply worlds/dark_souls_3/__init__.py:1426]: I'm confused, what space? This suggested change is removing a newline?

For everything within the `for ds3_world in ds3_worlds:` loop, other than the first line, I selected everything and pressed TAB to indent, and then I replaced `self` with `ds3_world`. There shouldn't be any other changes within the loop.

File: worlds/dark_souls_3/__init__.py [hunk 2]
```diff
@@ -1375,106 +1376,125 @@ def post_fill(self):
...
-            smooth_items([item for item in all_item_order if item.base_name in base_names])
-
-        if self.options.smooth_soul_items:
-            smooth_items([
-                item for item in all_item_order
-                if item.souls and item.classification != ItemClassification.progression
-            ])
+            # All DarkSouls3Items for this world that have been assigned anywhere, grouped by name
+            full_items_by_name: Dict[str, List[DarkSouls3Item]] = defaultdict(list)
+            for location in ds3_world.multiworld.get_filled_locations():
```
Comment (nex3) [inline worlds/dark_souls_3/__init__.py:1415]:
```suggestion
            for location in multiworld.get_filled_locations():
```

Similar change below

**Review by nex3 — APPROVED** [review]

File: worlds/dark_souls_3/__init__.py (hunk 1 above)
Comment (Exempt-Medic) [reply worlds/dark_souls_3/__init__.py:1426]: Hmm, I guess my text editor removed it for some reason. I copied the text and compared them and this line was different but looking again now, they are the same. This is resolved then

**Review by Exempt-Medic — APPROVED** [review]
Tested the changes by comparing generations before and after. Compared the lines before/after with indentation changed and all the changes made sense. The new code LGTM and the speedup with large multiworlds is significant. Just nex's comment's two changes left then.

**Comment by Mysteryem** [comment]
I think it would be best for a separate PR, but there's definitely some further improvements that could be made to reduce iterations through all locations in the multiworld, i.e., with a single pass, find all locations with a DS3 item belonging to a DS3 player with one of the smoothing options enabled.

Currently, `multiworld.filled_locations()` is iterated once per player and a second time for each player with weapon smoothing. Because `multiworld.filled_locations()` first iterates into a list and then that list is being iterated, this doubles the amount of iterating so ultimately results in iterating through all locations in the multiworld twice per player with smoothing enabled or four times when weapon smoothing enabled.

**Pull request merged**